lib/wasix/src/state/linker.rs

```diff
@@ -0,0 +1,1444 @@
+    DtorFuncWithInvalidSignature(String),
+
+    #[error("Destructor function {0} failed to run: {1}")]
+    DtorFunctionFailed(String, RuntimeError),
+
+    #[error("Failed to deallocate memory: {0}")]
+    DeallocationError(#[from] MemoryDeallocationError),
+}
+
+pub struct DylinkInfo {
+    pub mem_info: wasmparser::MemInfo,
+    pub needed: Vec<String>,
+}
+
+pub struct LinkedMainModule {
+    pub memory: Memory,
+    pub indirect_function_table: Table,
+    pub stack_low: u64,
+    pub stack_high: u64,
+}
+
+#[derive(Clone, Copy, PartialEq, Eq)]
+enum GlobalImportResolutionKind {
+    Mem,
+    Func,
+}
+
+impl GlobalImportResolutionKind {
+    fn to_unresolved(&self, name: String, global: Global) -> UnresolvedGlobal {
+        match self {
+            Self::Mem => UnresolvedGlobal::Mem(name, global),
+            Self::Func => UnresolvedGlobal::Func(name, global),
+        }
+    }
+}
+
+enum UnresolvedGlobal {
+    // A GOT.mem entry, should be resolved to an exported global from another module.
+    Mem(String, Global),
+    // A GOT.func entry, should be resolved to the address of an exported function
+    // from another module (e.g. an index into __indirect_function_table).
+    Func(String, Global),
+}
+
+#[derive(Default)]
+struct InProgressLinkState {
+    // All modules loaded in by this link operation. Used to remove all the modules
+    // in case a initializer function ends up failing to run.
+    module_handles: Vec<ModuleHandle>,
+
+    // List of all pending modules. We need this so we don't get stuck in an infinite
+    // loop when modules have circular dependencies.
+    pending_modules: Vec<PathBuf>,
+
+    // List of globals we didn't manage to resolve yet. As the final step, this list
+    // is iterated over and all globals filled in. If they still can't be resolved,
+    // the entire link operation fails. This only works for mutable globals, but clang
+    // appears to generate mutable globals for both GOT.mem and GOT.func.
+    // The list contains the import types (name + global type), as well as the actual
+    // (uninitialized) global we created for it.
+    unresolved_globals: Vec<UnresolvedGlobal>,
+
+    // List of modules for which init functions should be run. This needs to happen
+    // as the last step, since the init functions may access stub globals or functions
+    // which can't be resolved until the entire module tree is loaded in.
+    init_callbacks: Vec<Box<dyn (FnOnce(&mut StoreMut) -> Result<(), LinkError>) + Send + Sync>>,
+}
+
+enum MainInstanceState {
+    // Still in the process of linking the main module
+    Uninitialized,
+    // Main module linked, now waiting for its instance
+    Pending(InProgressLinkState),
+    // Main module linked and initialized
+    Initialized(Instance),
+}
+
+/// The linker is responsible for loading and linking dynamic modules at runtime,
+/// and managing the shared memory and indirect function table.
+#[derive(Clone)]
```

Comment:
Can we put an explanation for the flow and functionality somewhere, maybe here on the linker or just in the module doc comments?

Reply:
That's probably a very good idea, if I can figure out the exact flow myself 😄